### PR TITLE
RoR - Account Activation & Password Reset PR

### PR DESCRIPTION
**Pull request by sasaranAndrei**
This Pull Request is made for the sixth assignment of **Wofpack Digital RoR Internship**.

Made changes include:

- Add Account Activation via email
- Add Forgot Password / Password Reset functionality via email
- Add RSpec tests for unauthorized and unactivated users
- Style flash messages
- Deploy on **Heroku** -> [here ](https://eurekacaffe.herokuapp.com/)(with images from **Cloudinary**)

Login credentials:
email: 'test@eurekacaffe.com', password: 'testpassword'.

Sign-up functionality is not working properly since **Mailgun** can send emails just to the _Authorized Recipients_.

### CODE REVIEW

File: app/assets/stylesheets/_variables.scss [hunk 1]
```diff
@@ -1,3 +1,4 @@
 $site-color: rgb(255, 102, 0);
+$site-color-darker :rgb(165, 66, 0);
```
Comment (Sunzi2598) [inline app/assets/stylesheets/_variables.scss]: modify here to be like $site-color-darker: rgb(165, 66, 0);

File: app/assets/stylesheets/custom.scss [hunk 2]
```diff
@@ -38,3 +40,27 @@ p {
     color: red;
   }
 }
```
Comment (Sunzi2598) [inline app/assets/stylesheets/custom.scss:48]: add an empty line after every class to increase readability

File: app/controllers/account_activations_controller.rb [hunk 3]
```diff
@@ -0,0 +1,16 @@
+class AccountActivationsController < ApplicationController
+  def edit
+    user = User.find_by(email: params[:email])
+    if user&.unactivated? && user.authenticated?(:activation, params[:id])
```
Comment (Sunzi2598) [inline app/controllers/account_activations_controller.rb]: add an empty line before if also to increase readability

File: app/controllers/password_resets_controller.rb [hunk 4]
```diff
@@ -0,0 +1,60 @@
+class PasswordResetsController < ApplicationController
+  before_action :get_user, only: %i[edit update]
+  before_action :valid_user, only: %i[edit update]
+  before_action :check_expiration, only: %i[edit update] 
+
+  def new; end
+
+  def create
+    @user = User.find_by(email: params[:password_reset][:email].downcase)
+    if @user
```
Comment (Sunzi2598) [inline app/controllers/password_resets_controller.rb:11]: same here

File: app/controllers/sessions_controller.rb [hunk 5]
```diff
@@ -5,17 +5,27 @@ def create
     session_params = params[:session]
     user = User.find_by(email: session_params[:email].downcase)
     if user&.authenticate(session_params[:password])
```
Comment (Sunzi2598) [inline app/controllers/sessions_controller.rb:8]: same here

File: app/controllers/users_controller.rb [hunk 6]
```diff
@@ -11,9 +10,10 @@ def new
   def create
     @user = User.new(user_params)
     if @user.save
```
Comment (Sunzi2598) [inline app/controllers/users_controller.rb:13]: same here

File: app/assets/stylesheets/custom.scss [hunk 7]
```diff
@@ -38,3 +40,27 @@ p {
...
+.alert {
+  text-align: center;
+  border: 2px black solid;
+  color: black;
+}
```
Comment (andreeaiulia13) [inline app/assets/stylesheets/custom.scss:54]: Please add an empty line between each class style in this css file.

File: app/views/user_mailer/account_activation.html.haml [hunk 8]
```diff
@@ -0,0 +1,4 @@
+%h1 Eureka Caffe
```
Comment (Sunzi2598) [inline app/views/user_mailer/account_activation.html.haml:1]: please add empty lines to increase readability

File: app/controllers/account_activations_controller.rb (hunk 3 above)
Comment (andreeaiulia13) [inline app/controllers/account_activations_controller.rb]: I think that for better readability you could add an empty line between user and if

File: app/views/user_mailer/account_activation.text.haml [hunk 9]
```diff
@@ -0,0 +1,3 @@
+= "Hi #{ @user.name },"
```
Comment (Sunzi2598) [inline app/views/user_mailer/account_activation.text.haml:1]: same here

File: app/views/user_mailer/password_reset.html.haml [hunk 10]
```diff
@@ -0,0 +1,5 @@
+= "Hi #{ @user.name },"
```
Comment (Sunzi2598) [inline app/views/user_mailer/password_reset.html.haml:1]: same here

File: app/views/user_mailer/password_reset.text.haml [hunk 11]
```diff
@@ -0,0 +1,5 @@
+= "Hi #{ @user.name },"
```
Comment (Sunzi2598) [inline app/views/user_mailer/password_reset.text.haml:1]: same here

**Review by Sunzi2598 — CHANGES_REQUESTED** [review]
Good job 👍! Just add more empty lines to increase your readability

File: app/controllers/account_activations_controller.rb [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+      user.activate!
+      log_in user
```
Comment (andreeaiulia13) [inline app/controllers/account_activations_controller.rb]: I've observed that you are calling a method either with () or without. Try to be consistent when calling methods.

File: app/controllers/password_resets_controller.rb (hunk 4 above)
Comment (andreeaiulia13) [inline app/controllers/password_resets_controller.rb:11]: An empty line would look nice between user and if

File: spec/fixtures/user_mailer/account_activation [hunk 13]
```diff
@@ -0,0 +1,3 @@
+UserMailer#account_activation
```
Comment (andreeaiulia13) [inline spec/fixtures/user_mailer/account_activation]: Do you still need this file? If not, please delete it

File: spec/fixtures/user_mailer/password_reset [hunk 14]
```diff
@@ -0,0 +1,3 @@
+UserMailer#password_reset
```
Comment (andreeaiulia13) [inline spec/fixtures/user_mailer/password_reset]: same

**Review by andreeaiulia13 — CHANGES_REQUESTED** [review]
Very well-written code! You only need to make a few changes

File: app/assets/stylesheets/users.scss [hunk 15]
```diff
@@ -23,21 +23,26 @@ aside {
...
 .gravatar {
   float: left;
   margin-right: 10px;
 }
+
 .gravatar_edit {
   margin-top: 15px; 
 }
+
 .row{
```
Comment (andreeaiulia13) [inline app/assets/stylesheets/users.scss]: Please add a space between .row and {

**Review by andreeaiulia13 — APPROVED** [review]
God job! :clap:

**Review by Sunzi2598 — APPROVED** [review]

File: app/models/user.rb [hunk 16]
```diff
@@ -26,13 +27,52 @@ def remember!
...
+
+  def unactivated?
+    !activated?
+  end
+
+  private
+
+  def downcase_email
+    self.email = email.downcase
+  end
```
Comment (flaviuvr) [inline app/models/user.rb]: I think you could simplify this by writing just
`email.downcase!`

**Review by flaviuvr — COMMENTED** [review]
Well done🙌

File: app/models/user.rb (hunk 16 above)
Comment (sasaranAndrei) [reply app/models/user.rb]: Nice one! ✌️

File: app/controllers/account_activations_controller.rb [hunk 17]
```diff
@@ -0,0 +1,17 @@
+class AccountActivationsController < ApplicationController
+  def edit
+    user = User.find_by(email: params[:email])
+    
+    if user&.unactivated? && user.authenticated?(:activation, params[:id])
```
Comment (VictorMotogna) [inline app/controllers/account_activations_controller.rb]: I think you'll need to add `user&.authenticated` here too

File: app/controllers/password_resets_controller.rb [hunk 18]
```diff
@@ -0,0 +1,61 @@
+class PasswordResetsController < ApplicationController
+  before_action :get_user, only: %i[edit update]
+  before_action :valid_user, only: %i[edit update]
+  before_action :check_expiration, only: %i[edit update]
```
Comment (vaneaclaudia) [inline app/controllers/password_resets_controller.rb:4]: I think the naming of the methods are not accurate. For me `valid_user` would check if the user is valid and would return true/false (the method then would have a `?` ). Try to organise the code and name the methods so that it's clear enough what they are doing

File: app/controllers/sessions_controller.rb [hunk 19]
```diff
@@ -4,18 +4,29 @@ def new; end
...
+        redirect_to(forwarding_url || user)  
+      else
+        flash[:warning] = 'Account not activated. Please check your email for the activation link'
+        
+        redirect_to(root_url)
+      end
     else
+      flash[:danger] = 'Invalid email / password combination'
       render('new')
     end
```
Comment (vaneaclaudia) [inline app/controllers/sessions_controller.rb:13]: the method is too long, hard to follow the if else branches. Try to split up the logic a bit

File: app/models/user.rb [hunk 20]
```diff
@@ -26,13 +27,52 @@ def remember!
...
+    BCrypt::Password.new(digest).is_password?(token)
   end
 
   def forget!
     update_attribute(:remember_digest, nil)
   end
+
+  def activate!
+    update_attribute(:activated, true)
+    update_attribute(:activated_at, Time.zone.now)
```
Comment (vaneaclaudia) [inline app/models/user.rb]: you could have both updates with one transaction https://apidock.com/rails/ActiveRecord/Base/update/class

File: app/models/user.rb [hunk 21]
```diff
@@ -26,13 +27,52 @@ def remember!
...
+  end
+
+  def send_activation_email
+    UserMailer.account_activation(self).deliver_now
+  end
+
+  def create_reset_digest!
+    self.reset_token = User.new_token
+    update_attribute(:reset_digest, User.digest(reset_token))
+    update_attribute(:reset_sent_at, Time.zone.now)
```
Comment (vaneaclaudia) [inline app/models/user.rb]: same remark regarding update

File: spec/requests/access_site_spec.rb [hunk 22]
```diff
@@ -0,0 +1,21 @@
+require 'rails_helper'
+
+RSpec.describe 'Access site', type: :request do
+  
+  describe 'GET /login' do
+    let!(:user) { User.create(name: 'Name', email: 'email@domain.com', password: '123456', password_confirmation: '123456') }
```
Comment (vaneaclaudia) [inline spec/requests/access_site_spec.rb]: you could use `let` instead of `let!` since the `subject` is calling user.email so you would have the user in your context

File: spec/requests/access_site_spec.rb [hunk 23]
```diff
@@ -0,0 +1,21 @@
+require 'rails_helper'
+
+RSpec.describe 'Access site', type: :request do
```
Comment (vaneaclaudia) [inline spec/requests/access_site_spec.rb]: in Rspec.describe you should have the class that you are testing, not a description.

File: spec/requests/add_to_cart_spec.rb [hunk 24]
```diff
@@ -1,9 +1,21 @@
...
-    get '/'
-    
-    expect(response).to have_http_status(:success)
+RSpec.describe 'Add to cart', type: :request do
+
+  describe 'POST products#add_to_cart' do
+    let!(:product) { Product.create(title: 'title', description: 'Desserts', price: 5) }
+
+    subject do
+      product.update_attribute(:id, 999)
```
Comment (vaneaclaudia) [inline spec/requests/add_to_cart_spec.rb]: this should go in a before block

**Review by vaneaclaudia — CHANGES_REQUESTED** [review]

File: app/controllers/account_activations_controller.rb (hunk 17 above)
Comment (sasaranAndrei) [reply app/controllers/account_activations_controller.rb]: The default condition was `if user && user.unactivated? && user.authenticated?`. 
I tried to use `&.` operator for both conditions with something like 
`if user&.unactivated?&.authenticated?`  but I found out that this condition will be translated to 
`if user && user.unactivated? && user.unactivated.authenticated?` and this is not the desired behavior.
Also, I used `user&.unactivated? && user&.authenticated?` and it double-checked `user`, so I considered it’s better to avoid this.
What should I do?

File: app/models/product.rb [hunk 25]
```diff
@@ -2,7 +2,7 @@ class Product < ApplicationRecord
   validates :title, presence: true, uniqueness: true, length: { maximum: 18, too_long: "Title of a product can't be longer than %{count} characters." }
   validates :description, presence: true, inclusion: { in: %W(#{ "Main courses" } #{ "Second courses" } Soups Entrees Salads Desserts) }
   validates :price, presence: true, numericality: { greater_than: 0 }
-  validates :picture, presence: true
+  # validates :picture, presence: true
```
Comment (AdrianFlorian) [inline app/models/product.rb]: is this a mistake?

**Review by AdrianFlorian — CHANGES_REQUESTED** [review]
Nice job! 👏 just one very small comment :-)

**Comment by AdrianFlorian** [comment]
@sasaranAndrei nu merge signup-ul :(

**Comment by sasaranAndrei** [comment]
Am modificat descrierea PR-ului, legat de partea cu **Log in** & **Sign Up**.

File: app/controllers/password_resets_controller.rb [hunk 26]
```diff
@@ -0,0 +1,65 @@
...
+    elsif @user.update(user_params)
+      forward(@user)
+    else
+      render('edit')
+    end
+  end
+
+  private
+
+  def forward(user)
```
Comment (VictorMotogna) [inline app/controllers/password_resets_controller.rb]: this method naming is not very self-explanatory. you should try to rename it

File: spec/factories.rb [hunk 27]
```diff
@@ -0,0 +1,22 @@
...
+  
+  factory :user do
+    name { FactoryHelpers.random_string }
+    email "#{ FactoryHelpers.random_string }@gmail.com"
+    password '123456'
+    password_confirmation '123456'
+  end
+
+  factory :product do
+    id 1000
```
Comment (VictorMotogna) [inline spec/factories.rb]: I don't think you need to pass id to factory

File: spec/factories.rb [hunk 28]
```diff
@@ -0,0 +1,22 @@
+require 'factory_helpers'
+
+FactoryBot.define do
+  
+  factory :user do
+    name { FactoryHelpers.random_string }
```
Comment (VictorMotogna) [inline spec/factories.rb]: you should decide if you want to use everywhere:
```ruby
field { value }
# or
field value
```

I would recommend using the first one, adding `{ }` to all values

File: spec/factory_helpers.rb [hunk 29]
```diff
@@ -0,0 +1,8 @@
+module FactoryHelpers
```
Comment (VictorMotogna) [inline spec/factory_helpers.rb]: a good alternative for this would be using [Faker](https://github.com/faker-ruby/faker)

File: spec/requests/products_controller_spec.rb [hunk 30]
```diff
@@ -0,0 +1,18 @@
...
+RSpec.describe ProductsController, type: :request do
+
+  describe 'POST products#add_to_cart' do
+    before(:all) do
+      product = create(:product)
+
+      post add_to_cart_product_path(product.id)
+    end
+
+    context 'unauthorized user' do
```
Comment (VictorMotogna) [inline spec/requests/products_controller_spec.rb:8]: you should also add a context for a successful test, not just for unauthorized user. also, in that case you should structure:

```ruby
describe 'POST whatever' do
  context 'unauthorized user'
    before do
      # create the product, do any operations you need
    end

    it 'asks user for login' do
      expect..
    end
  end

  context 'successful add to cart'
    before do
      # create the product, do any operations you need
    end

    it 'successfully adds product to cart' do
      expect..
    end
  end
end
```

File: spec/requests/sessions_controller_spec.rb [hunk 31]
```diff
@@ -0,0 +1,20 @@
...
+RSpec.describe SessionsController, type: :request do
+
+  describe 'POST sessions#create' do
+    before(:all) do
+      user = create(:user)
+      
+      post login_path, params: { session: { email: user.email, password: user.password } }
+    end
+  
+    context 'unactivated user' do
```
Comment (VictorMotogna) [inline spec/requests/sessions_controller_spec.rb:9]: same logic here, you should add a test for a successful operation also

File: spec/factories.rb [hunk 32]
```diff
@@ -0,0 +1,19 @@
+FactoryBot.define do
+  
+  factory :user do
+    name 'Test Name'
```
Comment (VictorMotogna) [inline spec/factories.rb:4]: although not required, if you want you can also take a look over [Faker](https://github.com/faker-ruby/faker)

**Review by VictorMotogna — APPROVED** [review]

**Review by VictorMotogna — APPROVED** [review]

File: lib/tasks/factory_bot.rake [hunk 33]
```diff
@@ -0,0 +1,13 @@
+namespace :factory_bot do
+  desc "Verify that all FactoryBot factories are valid"
```
Comment (vaneaclaudia) [inline lib/tasks/factory_bot.rake:2]: why do you need this task?

**Review by vaneaclaudia — APPROVED** [review]

File: lib/tasks/factory_bot.rake (hunk 33 above)
Comment (sasaranAndrei) [reply lib/tasks/factory_bot.rake:2]: I had some issues with `database_cleaner` and `factory_bot` and I tried this solution from [here](https://semaphoreci.com/community/tutorials/working-effectively-with-data-factories-using-factorygirl).
Can I delete this file since the task was executed?

Comment (vaneaclaudia) [reply lib/tasks/factory_bot.rake:2]: If you are not using the task anymore I guess you can remove it.
